### ISSUE TITLE
High throughput for load phase, very low throughput for workload phase

### ISSUE THREAD

**Issue by seybi87** [brianfrankcooper/YCSB#646]
Hi YCSB guys,

I've run the YCSB in the Cloud against Couchbase and Cassandra and experienced for the load phase a throughput of ~15.000 ops/s.

However for the default workloadA/workloadB I got really poor results: Cochbase ~ 100 ops/s and Cassandra ~500 ops/s

Benchmark setup:
Cloud: private Openstack 
DB VM: 4 cores, 8GB RAM, 50 GB Disk HDD
YCSB VM: 2 Cores, 4GB RAM, 20 GB disk
Records: 10.000.000
Operations: 10.000.000
Distribution: zipfian (also used latest without significant changes)
YCSB fork: [YCSB](https://github.com/seybi87/YCSB)

Do you have any axplanation for this low throughput because I expected a much higher throughput for workloadA/B (based on the results in several publications/white papers using the YCSB to benchmark Couchbase/YCSB) 

Thanks in advance for your help!

Cheers,
Daniel

**Comment by kruthar** [comment]
Can you post your lad ad run commands. Was there any significant output during the workloads? Also, can you post your final result output for the load and run phases?

**Comment by seybi87** [comment]
Hi,

find here the load command for the Couchbase benchmark:

`[WARN]  Running against a source checkout. In order to get our runtime dependencies we'll have to invoke Maven. Depending on the state of your system, this may take ~30-45 seconds
[DEBUG]  Running 'mvn -pl com.yahoo.ycsb:couchbase2-binding -am package -DskipTests dependency:build-classpath -DincludeScope=compile -Dmdep.outputFilterFile=true'
java -cp /home/ubuntu/YCSB-master/couchbase2/conf:/home/ubuntu/YCSB-master/couchbase2/target/couchbase2-binding-0.6.0-SNAPSHOT.jar:/home/ubuntu/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.2.2/jackson-annotations-2.2.2.jar:/home/ubuntu/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.2.2/jackson-databind-2.2.2.jar:/home/ubuntu/.m2/repository/io/reactivex/rxjava/1.0.15/rxjava-1.0.15.jar:/home/ubuntu/.m2/repository/com/couchbase/client/core-io/1.2.2/core-io-1.2.2.jar:/home/ubuntu/YCSB-master/core/target/core-0.6.0-SNAPSHOT.jar:/home/ubuntu/.m2/repository/org/codehaus/jackson/jackson-core-asl/1.9.4/jackson-core-asl-1.9.4.jar:/home/ubuntu/.m2/repository/org/hdrhistogram/HdrHistogram/2.1.4/HdrHistogram-2.1.4.jar:/home/ubuntu/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.2.2/jackson-core-2.2.2.jar:/home/ubuntu/.m2/repository/com/couchbase/client/java-client/2.2.2/java-client-2.2.2.jar:/home/ubuntu/.m2/repository/org/codehaus/jackson/jackson-mapper-asl/1.9.4/jackson-mapper-asl-1.9.4.jar:/home/ubuntu/.m2/repository/org/slf4j/slf4j-api/1.6.4/slf4j-api-1.6.4.jar com.yahoo.ycsb.Client -db com.yahoo.ycsb.db.CouchbaseClient -s -P workloads/workloada -P couchbase2/conf/couchbase.properties -load
YCSB Client 0.1
Command line: -db com.yahoo.ycsb.db.CouchbaseClient -s -P workloads/workloada -P couchbase2/conf/couchbase.properties -load
Loading workload...`

And for running the workloadA:
`[WARN]  Running against a source checkout. In order to get our runtime dependencies we'll have to invoke Maven. Depending on the state of your system, this may take ~30-45 seconds
[DEBUG]  Running 'mvn -pl com.yahoo.ycsb:couchbase2-binding -am package -DskipTests dependency:build-classpath -DincludeScope=compile -Dmdep.outputFilterFile=true'
java -cp /home/ubuntu/YCSB-master/couchbase2/conf:/home/ubuntu/YCSB-master/couchbase2/target/couchbase2-binding-0.6.0-SNAPSHOT.jar:/home/ubuntu/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.2.2/jackson-annotations-2.2.2.jar:/home/ubuntu/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.2.2/jackson-databind-2.2.2.jar:/home/ubuntu/.m2/repository/io/reactivex/rxjava/1.0.15/rxjava-1.0.15.jar:/home/ubuntu/.m2/repository/com/couchbase/client/core-io/1.2.2/core-io-1.2.2.jar:/home/ubuntu/YCSB-master/core/target/core-0.6.0-SNAPSHOT.jar:/home/ubuntu/.m2/repository/org/codehaus/jackson/jackson-core-asl/1.9.4/jackson-core-asl-1.9.4.jar:/home/ubuntu/.m2/repository/org/hdrhistogram/HdrHistogram/2.1.4/HdrHistogram-2.1.4.jar:/home/ubuntu/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.2.2/jackson-core-2.2.2.jar:/home/ubuntu/.m2/repository/com/couchbase/client/java-client/2.2.2/java-client-2.2.2.jar:/home/ubuntu/.m2/repository/org/codehaus/jackson/jackson-mapper-asl/1.9.4/jackson-mapper-asl-1.9.4.jar:/home/ubuntu/.m2/repository/org/slf4j/slf4j-api/1.6.4/slf4j-api-1.6.4.jar com.yahoo.ycsb.Client -db com.yahoo.ycsb.db.CouchbaseClient -s -P workloads/workloada -P couchbase2/conf/couchbase.properties -t
YCSB Client 0.1
Command line: -db com.yahoo.ycsb.db.CouchbaseClient -s -P workloads/workloada -P couchbase2/conf/couchbase.properties -t
Loading workload...`

I've checked the output of the resulting files (stdout/errorout) but there were no significant errors/exceptions.
The same for the Cassandra output.

Cheers,
Daniel

**Comment by kruthar** [comment]
Is there anything significant in couchbase.properties? can you post the content of that file.

Also, have you altered the project in any significant way? We don't have a ```couchbase2``` binding.

Last thing - can you post the performance statistics from the end of the workload. It should show things like overall time taken, average latencies and return code counts.

**Comment by seybi87** [comment]
I did not use any specific properties in the couchbase.properties file:

`couchbase.url=IP_TO_COUCHABSE
couchbase.bucket=default
couchbase.password=
couchbase.checkFutures=true
couchbase.persistTo=0
couchbase.replicateTo=0
couchbase.json=true`

Yes I added the couchbase2 binding in my fork of the YCSB, using the new Couchabse Client 2.X.
However I also did the benchmarks with the original Couchbase binding as well (same results)

Finde here the performance statistics:
LOAD:
`[OVERALL], RunTime(ms), 365728.0
[OVERALL], Throughput(ops/sec), 27342.724647825708
[CLEANUP], Operations, 20.0
[CLEANUP], AverageLatency(us), 7106.95
[CLEANUP], MinLatency(us), 1466.0
[CLEANUP], MaxLatency(us), 30447.0
[CLEANUP], 95thPercentileLatency(us), 15495.0
[CLEANUP], 99thPercentileLatency(us), 30447.0
[INSERT], Operations, 1.0E7
[INSERT], AverageLatency(us), 716.6713845
[INSERT], MinLatency(us), 174.0
[INSERT], MaxLatency(us), 127487.0
[INSERT], 95thPercentileLatency(us), 1194.0
[INSERT], 99thPercentileLatency(us), 2443.0
[INSERT], Return=OK, 10000000`

WorkloadA (only intermediate results):
`2016-03-02 14:30:09:700; 80; sec; 13292 operations; 171.3 current ops/sec; est completion in 16 hours 41 minutes [READ: Count=1624, Max=821247, Min=211, Avg=114736.14, 90=393983, 99=652287, 99.9=775679, 99.99=821247] [UPDATE: Count=88, Max=934399, Min=298, Avg=128362.9, 90=403455, 99=737279, 99.9=934399, 99.99=934399]

2016-03-02 14:30:19:700; 90; sec; 14949 operations; 165.7 current ops/sec; est completion in 16 hours 41 minutes [READ: Count=1570, Max=1925119, Min=210, Avg=121408.37, 90=375295, 99=886783, 99.9=1699839, 99.99=1925119] [UPDATE: Count=87, Max=1707007, Min=343, Avg=122345.4, 90=332031, 99=553983, 99.9=1707007, 99.99=1707007]

2016-03-02 14:30:29:700; 100; sec; 16443 operations; 149.4 current ops/sec; est completion in 16 hours 51 minutes [READ: Count=1417, Max=1179647, Min=200, Avg=131635.51, 90=420095, 99=798207, 99.9=1142783, 99.99=1179647] [UPDATE: Count=77, Max=1047039, Min=441, Avg=142450.65, 90=417023, 99=740351, 99.9=1047039, 99.99=1047039]

2016-03-02 14:30:39:700; 110; sec; 18213 operations; 177 current ops/sec; est completion in 16 hours 44 minutes [READ: Count=1699, Max=1381375, Min=217, Avg=114653.49, 90=387839, 99=726527, 99.9=979455, 99.99=1381375] [UPDATE: Count=72, Max=856575, Min=310, Avg=110919.58, 90=384767, 99=740863, 99.9=856575, 99.99=856575]

2016-03-02 14:30:49:700; 120; sec; 19689 operations; 147.6 current ops/sec; est completion in 16 hours 53 minutes [READ: Count=1395, Max=1564671, Min=202, Avg=135217.15, 90=428799, 99=977919, 99.9=1517567, 99.99=1564671] [UPDATE: Count=80, Max=1421311, Min=240, Avg=133606.48, 90=386559, 99=669695, 99.9=1421311, 99.99=1421311]`

Cheers Daniel

**Comment by kruthar** [comment]
Interesting, your inserts are super fast, and the reads are very slow as you say. From what you posted nothing is jumping out at me.

Can you confirm that all 10,000,000 records are in Couchbase after the load? What is the average doc size? Total data set size?

For Couchbase, how much memory did you allocate to it? One possibility is that Couchbase doesn't have enough memory. Inserts may not be effected by that, but if Couchbase doesn't have much memory to cache records, and it is constantly having to go to disk then that can be an issue. Not all of the reads are taking a long time, there is a min read time in your output of just of just 200us which is pretty fast.

Take a look at machine statistics while running the transaction phase, what does the cpu and mem look like on the Couchbase machines and on the YCSB machines. Couchbase also has a good metrics interface you can take a look it, anything look strange or toping out there?

**Comment by seybi87** [comment]
Yes this super fast insert and very low read/update throughput really keeps me struggeling...

I've always checked the inserted records count and it's 10.000.000.

I allocated to Couchbase 5.5GB of RAM (5GB to data and 0.5GB to index)
The bucket is configured to use 5GB, value ejection, disk I/O priority low (highhas shown the same results).

I have also monitored the YCSB client with ganlia and Couchbase with the built-in metric interface and Ganglia.
The client shows a ~70% CPU usage at the load however at the read/upadate only ~15%
Couchbase RAM was nearly fully used (due to caching) however the CPU was not heavy used ~40% for the read/update.

I have also benchmarked the disk throughput on the VMs and the throughput is as high as expected (also proven by the high load throughput)

**Comment by risdenk** [comment]
If you change couchbase.persistTo=0 to 1 does that significantly affect the load?

**Comment by kruthar** [comment]
OK, so, please double check my math here. You are using the default configurations which are:

```
fieldcount=10
fieldlength=100
```

So, that means an estimate for your total data size is 10,000,000 records * 10 fields * 100 bytes = 10 GB.

You are using ```persistTo=0``` and ```replicateTo=0``` so your inserts will be very fast. But on the reads, only half of your dataset can fit in Couchbase cache memory, so 50% of your reads are having to go all the way to disk.

Can you test this theory by trying your load and transaction phases again but with only 1/3 of the records? or 1/4 of the records, just to be sure that Couchbase should have an easy time with keeping everything in memory. Make sure that you empty or recreate the bucket first.

**Comment by seybi87** [comment]
Yes I am using the default configuration for fieldcount and fieldlength which results in a total data size of ~10GB for Couchbase.

I have run the load phase with the `persistTo=1` and I got ~ `20.4 current ops/sec;` and a lot of timeout. Of course a synchronous write to the disk will decrease the write performance. 

Further I've run the load and WorkloadA with 1.000.000 of records (so they will fit completely in memory) with the same write throughput (~20.000) and also a high read/update throughput (25.000).

However I'd like to benchmark the databases with the focus on large datasets which will not fit completely in memory.
And I did not assume such a low throughput as soon as datasets are fetched from disk.

Did you run similiar benchmarks   earlier?

**Comment by kruthar** [comment]
So, from a YCSB perspective I would call your findings a success. You used YCSB to test two main scenarios: 1) a recordset that fits in memory and 2) a recordset that does not fit in memory. What you found is that a good throughput when you could fit the set in memory and a decisive performance decrease when you cannot.

This was not so much a finding (you already knew this would happen), but it was verification of your hypothesis, that performance would take a hit if Couchbase could not fit the entire working set in memory.

I don't actually see any "problems" with any of this. So, at this point I think you have to take a step back and figure out the exact question you want to answer. Your last comment:

>However I'd like to benchmark the databases with the focus on large datasets which will not fit completely in memory.
And I did not assume such a low throughput as soon as datasets are fetched from disk.

You may not have expected such a low throughput under your high stress scenario, but you have benchmarking metrics that show a certain performance under certain conditions. There doesn't seem to be a whole lot more on the YCSB front that you can tweak or change. If you expect better performance you have to go to your database system and find the optimal configuration there.

**Comment by seybi87** [comment]
Thanks a lot for your helpful comments!

I already used the YCSB  around 1 ago with a similar setup. However I used the fork of [Thumbtack ](https://github.com/thumbtack-technology/ycsb) where I experienced a much higher throughput for 10.000.000 records/operations. Yet I did not recognize they the fork modified the record size to only 100 bytes instead of the 1KB. This explains the different results with my current setup.
